### PR TITLE
[Fix][FSDP]fixed padding size of input tensor for reduce scatter

### PR DESCRIPTION
**Pull request by tmarkstrum**
## What does this PR do?
Fixes # 904

a. pad the input tensor for reduce scatter operation to the size of the process group used for the reduce scatter operation.
b. set the reduce scatter process group to be parameter value of process_group_reduce_scatter when it is specified as a specific ProcessGroup.

## Before submitting

- [ ] Did you have fun?
  - Make sure you had fun coding 🙃
- [ ] Did you read the [contributor guideline](https://github.com/facebookresearch/fairscale/blob/main/CONTRIBUTING.md)?
- [ ] Was this discussed/approved via a Github issue? (no need for typos, doc improvements)
  - [ ] N/A
- [ ] Did you make sure to update the docs?
  - [ ] N/A
- [ ] Did you write any new necessary tests?
  - [ ] N/A
- [ ] Did you update the [changelog](https://github.com/facebookresearch/fairscale/blob/main/CHANGELOG.md)? (if needed)
  - [ ] N/A

## PR review
Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

### CODE REVIEW

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 1]
```diff
@@ -298,7 +298,7 @@ def __init__(
         self,
         module: nn.Module,
         process_group: Optional[ProcessGroup] = None,
-        process_group_reduce_scatter: Union[ProcessGroup, ProcessGroupName] = ProcessGroupName.reduce_scatter,
+        process_group_reduce_scatter: Any = ProcessGroupName.reduce_scatter,
```
Comment (tmarkstrum) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:301]: It has to be Any. Or else it would not pass mypy type check with the error message. 
" Incompatible types in assignment (expression has type "ProcessGroupName",
variable has type "ProcessGroup")
    ...             self.process_group_reduce_scatter = process_group_reduce_... "
I added extra type check in the code below to ensure correct type is passed in.

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 2]
```diff
@@ -1617,7 +1629,7 @@ def _post_backward_hook(self, param: Parameter, *unused: Any) -> None:
                 # unsharded gradients allocated; one for a pending reduction, and one for gradient computation.
                 param.grad = None
                 callback_fn = functools.partial(self._post_reduction_hook, param)
-                grad_chunks = chunk_and_pad(grad, self.world_size)
+                grad_chunks = chunk_and_pad(grad, self.process_group_reduce_scatter.size())
```
Comment (anj-s) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1632]: Can you add a test for this?

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py (hunk 1 above)
Comment (min-xu-ai) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:301]: Can you add a comment here please?

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 3]
```diff
@@ -323,12 +323,24 @@ def __init__(
...
+            if isinstance(process_group_reduce_scatter, ProcessGroup):
+                self.process_group_reduce_scatter = process_group_reduce_scatter
+            else:
+                if not hasattr(process_group_reduce_scatter, "allgather") and hasattr(
+                    process_group_reduce_scatter, "rank"
+                ):
+                    # Likely a dummy pg for unit test
+                    self.process_group_reduce_scatter = process_group_reduce_scatter
+                else:
+                    raise TypeError("upsupported type")
```
Comment (anj-s) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:343]: typo: "unsupported

Comment (min-xu-ai) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:343]:
```suggestion
                    raise TypeError("unsupported type for reduce_scatter process group")
```

**Review by min-xu-ai — APPROVED** [review]
nice!

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py [hunk 4]
```diff
@@ -323,12 +323,24 @@ def __init__(
         self.process_group = process_group or get_process_group_cached()
         # If ProcessGroupName.default is passed in, the reduce_scatter will use the same process group with
         # the rest of operations. The overlap feature in the backward propagation is disabled.
+        # If ProcessGroupName.reduce_scatter is passed in, the reduce_scatter use a seperate process group
```
Comment (anj-s) [inline fairscale/nn/data_parallel/fully_sharded_data_parallel.py:326]: I would move the comment to the if/else statement that it applies to.

**Comment by anj-s** [comment]
Can you also update the CHANGELOG?

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py (hunk 4 above)
Comment (tmarkstrum) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:326]: will do.

File: fairscale/nn/data_parallel/fully_sharded_data_parallel.py (hunk 2 above)
Comment (tmarkstrum) [reply fairscale/nn/data_parallel/fully_sharded_data_parallel.py:1632]: All the unit tests are still passing. Our unittest env only have 2 gpus. It is difficult to create default process group and reduce scatter process group with different number of ranks. By default, both default process group and reduce scatter process group are using all the ranks in dist. The size of ranks are the same.  #904 probably uses unconventional way to initiate the default process group.

**Review by anj-s — APPROVED** [review]
